### DISCUSSION TITLE
Error generating test sentences on multi-speaker

### DISCUSSION THREAD

**Discussion by Lizh-TW** [coqui-ai/TTS#760] [General Q&A] [answered]
I have read https://github.com/coqui-ai/TTS/discussions/496#discussion-3375296
Is that problem has solved ?
Please tell me how to solve this problem, thank you!!!
**ERROR**

   --> STEP: 70/77 -- GLOBAL_STEP: 850
     | > decoder_loss: 0.46232  (0.47110)
     | > postnet_loss: 0.51946  (0.54816)
     | > stopnet_loss: 0.69046  (0.87511)
     | > decoder_coarse_loss: 0.37904  (0.37771)
     | > decoder_ddc_loss: 0.00209  (0.00434)
     | > ga_loss: 0.00080  (0.00161)
     | > decoder_diff_spec_loss: 0.10055  (0.10126)
     | > postnet_diff_spec_loss: 0.23031  (0.24835)
     | > decoder_ssim_loss: 0.63622  (0.59308)
     | > postnet_ssim_loss: 0.64565  (0.60485)
     | > loss: 1.43838  (1.62037)
     | > align_error: 0.95027  (0.91372)
     | > grad_norm: 3.21526  (2.79398)
     | > step_time: 0.72460  (0.47089)
     | > loader_time: 0.00290  (0.00229)

 > DataLoader initialization
 | > Use phonemes: True
   | > phoneme language: zh-CN
 | > Number of instances : 50
 | > Max length sequence: 115
 | > Min length sequence: 18
 | > Avg length sequence: 59.58
 | > Num. instances discarded by max-min (max=180, min=6) seq limits: 0
 | > Batch group size: 0.

 > EVALUATION

 | > Synthesizing test sentences.
Building prefix dict from the default dictionary ...
Building prefix dict from the default dictionary ...
Loading model from cache /tmp/jieba.cache
Loading model from cache /tmp/jieba.cache
Loading model cost 0.488 seconds.
Loading model cost 0.488 seconds.
Prefix dict has been built successfully.
Prefix dict has been built successfully.
 > Phonemes: i|1| | | |e|r|4| | | |s|a|n|1| | | |s|ɪ|4| | | |w|u|3| | | |l|i|o|4
 ! Run is kept in ./aishell3-ddc-August-21-2021_10+45PM-0000000
Traceback (most recent call last):
  File "/workspace/TTS/TTS/trainer.py", line 848, in fit
    self._fit()
  File "/workspace/TTS/TTS/trainer.py", line 838, in _fit
    self.test_run()
  File "/workspace/TTS/TTS/trainer.py", line 815, in test_run
    figures, audios = self.model.test_run(self.ap)
  File "/workspace/TTS/TTS/tts/models/base_tts.py", line 271, in test_run
    do_trim_silence=False,
  File "/workspace/TTS/TTS/tts/utils/synthesis.py", line 259, in synthesis
    outputs = run_model_torch(model, text_inputs, speaker_id, style_mel, d_vector=d_vector)
  File "/workspace/TTS/TTS/tts/utils/synthesis.py", line 98, in run_model_torch
    "style_mel": style_mel,
  File "/opt/conda/lib/python3.7/site-packages/torch/autograd/grad_mode.py", line 28, in decorate_context
    return func(*args, **kwargs)
  File "/workspace/TTS/TTS/tts/models/tacotron2.py", line 199, in inference
    encoder_outputs = self._concat_speaker_embedding(encoder_outputs, embedded_speakers)
  File "/workspace/TTS/TTS/tts/models/base_tacotron.py", line 265, in _concat_speaker_embedding
    embedded_speakers_ = embedded_speakers.expand(outputs.size(0), outputs.size(1), -1)
RuntimeError: The expanded size of the tensor (28) must match the existing size (449) at non-singleton dimension 1.  Target sizes: [1, 28, -1].  Tensor sizes: [1, 449, 512]

**Config**
{
    "model": "Tacotron2",
    "datasets": [
        {
            "name": "aishell3",
            "path": "./dataset/aishell_3",
            "meta_file_train": "metadata.csv",
            "meta_file_val": null
        }
    ],
    "audio": {
        "fft_size": 1024,
        "win_length": 1024,
        "hop_length": 256,
        "frame_length_ms": null,
        "frame_shift_ms": null,
        "sample_rate": 22050,
        "preemphasis": 0.0,
        "ref_level_db": 20,
        "do_trim_silence": true,
        "trim_db": 60,
        "power": 1.5,
        "griffin_lim_iters": 60,
        "num_mels": 80,
        "mel_fmin": 50.0,
        "mel_fmax": 7600.0,
        "spec_gain": 1,
        "signal_norm": true,
        "min_level_db": -100,
        "symmetric_norm": true,
        "max_norm": 4.0,
        "clip_norm": true,
        "stats_path": "scale_stats.npy"
    },
    "gst":{
        "gst_embedding_dim": 256,
        "gst_num_heads": 4,
        "gst_num_style_tokens": 10
    },
    "distributed_backend": "gloo",
    "distributed_url": "tcp:\/\/localhost:54321",
    "run_name": "aishell3-ddc",
    "run_description": "tacotron2 with double decoder consistency.",
    "batch_size": 32,
    "eval_batch_size": 32,
    "mixed_precision": false,
    "loss_masking": true,
    "decoder_loss_alpha": 0.25,
    "postnet_loss_alpha": 0.25,
    "postnet_diff_spec_alpha": 0.25,
    "decoder_diff_spec_alpha": 0.25,
    "decoder_ssim_alpha": 0.25,
    "postnet_ssim_alpha": 0.25,
    "ga_alpha": 5.0,
    "stopnet_pos_weight": 15.0,
    "run_eval": true,
    "test_delay_epochs": 10,
    "test_sentences_file": null,
    "max_decoder_steps": 1000,
    "noam_schedule": true,
    "grad_clip": 0.05,
    "epochs": 1000,
    "lr": 0.001,
    "wd": 1e-06,
    "warmup_steps": 4000,
    "memory_size": -1,
    "prenet_type": "original",
    "prenet_dropout": true,
    "attention_type": "original",
    "location_attn": true,
    "double_decoder_consistency": true,
    "ddc_r": 7,
    "attention_norm": "sigmoid",
    "r": 7,
    "gradual_training": [[0, 7, 64], [10000, 4, 32], [50000, 3, 32], [100000, 2, 32]],
    "stopnet": true,
    "separate_stopnet": true,
    "print_step": 25,
    "tb_plot_step": 100,
    "print_eval": false,
    "save_step": 10000,
    "checkpoint": true,
    "text_cleaner": "basic_cleaners",
    "num_loader_workers": 4,
    "num_val_loader_workers": 4,
    "batch_group_size": 4,
    "min_seq_len": 6,
    "max_seq_len": 180,
    "compute_input_seq_cache": true,
    "output_path": "./",
    "phoneme_cache_path": "./phoneme_cache",
    "use_phonemes": true,
    "phoneme_language":"zh-CN",
    "use_speaker_embedding": true,
    "use_d_vector_file": true,
    "d_vector_file":"./embedding/speakers.json",
    "d_vector_dim": 512,
    "test_sentences": [
        "一 二 三 四 五 六",
        "今 天 早 上 會 下 雨",
        "明 天 是 禮 拜 六",
        "我 正 在 嘗 試 說 話",
        "黑 化 肥 會 揮 發"
    ]
}

**Comment by Lizh-TW** [comment]
Is some difficulty with d_vector?

I have read https://edresson.github.io/SC-GlowTTS/ and try to reproduce it.

But when I use the d vector file to train, I always get a problem of generating test sentences, whether I use Glow_tts、Tacotron 2 or Tacotron, both in  English and Chinese.

**Comment by erogol** [comment]
@Lizh-TW I didn't get what you mean exactly. Did you try SC-GlowTTS and didn't work?

**Reply by Lizh-TW** [reply]
@erogol Hi.
I tried, but generating test sentences also get an error.
> EVALUATION

warning: audio amplitude out of range, auto clipped.
 | > Synthesizing test sentences.
Building prefix dict from the default dictionary ...
Building prefix dict from the default dictionary ...
Loading model from cache /tmp/jieba.cache
Loading model from cache /tmp/jieba.cache
Loading model cost 0.571 seconds.
Loading model cost 0.571 seconds.
Prefix dict has been built successfully.
Prefix dict has been built successfully.
 > Phonemes: d|z|a|ʌ|3|a|n|1| |，| |n|i|3|x|a|ʌ|3| |i|a|5| |!
 ! Run is kept in ./glow_tts_aishell3_test-August-31-2021_08+18AM-0000000
Traceback (most recent call last):
  File "/workspace/TTS/TTS/trainer.py", line 848, in fit
    self._fit()
  File "/workspace/TTS/TTS/trainer.py", line 838, in _fit
    self.test_run()
  File "/workspace/TTS/TTS/trainer.py", line 815, in test_run
    figures, audios = self.model.test_run(self.ap)
  File "/opt/conda/lib/python3.7/site-packages/torch/autograd/grad_mode.py", line 28, in decorate_context
    return func(*args, **kwargs)
  File "/workspace/TTS/TTS/tts/models/glow_tts.py", line 401, in test_run
    do_trim_silence=False,
  File "/workspace/TTS/TTS/tts/utils/synthesis.py", line 259, in synthesis
    outputs = run_model_torch(model, text_inputs, speaker_id, style_mel, d_vector=d_vector)
  File "/workspace/TTS/TTS/tts/utils/synthesis.py", line 98, in run_model_torch
    "style_mel": style_mel,
  File "/opt/conda/lib/python3.7/site-packages/torch/autograd/grad_mode.py", line 28, in decorate_context
    return func(*args, **kwargs)
  File "/workspace/TTS/TTS/tts/models/glow_tts.py", line 290, in inference
    o_mean, o_log_scale, o_dur_log, x_mask = self.encoder(x, x_lengths, g=g)
  File "/opt/conda/lib/python3.7/site-packages/torch/nn/modules/module.py", line 1051, in _call_impl
    return forward_call(*input, **kwargs)
  File "/workspace/TTS/TTS/tts/layers/glow_tts/encoder.py", line 178, in forward
    logw = self.duration_predictor(x_dp, x_mask)
  File "/opt/conda/lib/python3.7/site-packages/torch/nn/modules/module.py", line 1051, in _call_impl
    return forward_call(*input, **kwargs)
  File "/workspace/TTS/TTS/tts/layers/glow_tts/duration_predictor.py", line 48, in forward
    x = self.conv_1(x * x_mask)
  File "/opt/conda/lib/python3.7/site-packages/torch/nn/modules/module.py", line 1051, in _call_impl
    return forward_call(*input, **kwargs)
  File "/opt/conda/lib/python3.7/site-packages/torch/nn/modules/conv.py", line 298, in forward
    return self._conv_forward(input, self.weight, self.bias)
  File "/opt/conda/lib/python3.7/site-packages/torch/nn/modules/conv.py", line 295, in _conv_forward
    self.padding, self.dilation, self.groups)
RuntimeError: Given groups=1, weight of size [256, 704, 3], expected input[1, 192, 37] to have 704 channels, but got 192 channels instead
**config**

{
    "model": "glow_tts",
    "run_name": "glow_tts_aishell3_test",
    "run_description": "glow_tts_aishell3_test.",

    // AUDIO PARAMETERS
    // AUDIO PARAMETERS
    "audio":{
        // Audio processing parameters
        "num_mels": 80,         // size of the mel spec frame. 
        "fft_size": 1024,       // number of stft frequency levels. Size of the linear spectogram frame.
        "sample_rate": 22050, //22050,   // DATASET-RELATED: wav sample-rate. If different than the original data, it is resampled.
        "win_length": 1024,     // stft window length in ms.
        "hop_length": 256,      // stft window hop-lengh in ms.
        "frame_length_ms": null,  // stft window length in ms.If null, 'win_length' is used.
        "frame_shift_ms": null,   // stft window hop-lengh in ms. If null, 'hop_length' is used.
        "preemphasis": 0.98,    // pre-emphasis to reduce spec noise and make it more structured. If 0.0, no -pre-emphasis.
        "min_level_db": -100,   // normalization range
        "ref_level_db": 20,     // reference level db, theoretically 20db is the sound of air.
        "power": 1.5,           // value to sharpen wav signals after GL algorithm.
        "griffin_lim_iters": 60,// #griffin-lim iterations. 30-60 is a good range. Larger the value, slower the generation.
        "stft_pad_mode": "reflect",
        // Normalization parameters
        "signal_norm": true,    // normalize the spec values in range [0, 1]
        "symmetric_norm": true, // move normalization to range [-1, 1]
        "max_norm": 4.0,          // scale normalization to range [-max_norm, max_norm] or [0, max_norm]
        "clip_norm": true,      // clip normalized values into the range.
        "mel_fmin": 0.0,         // minimum freq level for mel-spec. ~50 for male and ~95 for female voices. Tune for dataset!!
        "mel_fmax": 8000.0,        // maximum freq level for mel-spec. Tune for dataset!!
        "spec_gain": 20.0, 
        "do_trim_silence": false,  // enable trimming of slience of audio as you load it. LJspeech (false), TWEB (false), Nancy (true)
        "trim_db": 60          // threshold for timming silence. Set this according to your dataset.
    },
    // VOCABULARY PARAMETERS
    // if custom character set is not defined,
    // default set in symbols.py is used
    //"characters":{
    //    "pad": "_",
    //    "eos": "&",
    //    "bos": "*",
    //    "characters": "ABCDEFGHIJKLMNOPQRSTUVWXYZÇÃÀÁÂÊÉÍÓÔÕÚÛabcdefghijklmnopqrstuvwxyzçãàáâêéíóôõúû!(),-.:;? ",
    //    "punctuations":"!'(),-.:;? ",
    //    "phonemes":"iyɨʉɯuɪʏʊeøɘəɵɤoɛœɜɞʌɔæɐaɶɑɒᵻʘɓǀɗǃʄǂɠǁʛpbtdʈɖcɟkɡqɢʔɴŋɲɳnɱmʙrʀⱱɾɽɸβfvθðszʃʒʂʐçʝxɣχʁħʕhɦɬɮʋɹɻjɰlɭʎʟˈˌːˑʍwɥʜʢʡɕʑɺɧɚ˞ɫ'̃' "
    //},

    "add_blank": true, // if true add a new token after each token of the sentence. This increases the size of the input sequence, but has considerably improved the prosody of the GlowTTS model.

    // DISTRIBUTED TRAINING
    "distributed":{
        "backend": "nccl",
        "url": "tcp:\/\/localhost:54321"
    },

    "reinit_layers": [],    // give a list of layer names to restore from the given checkpoint. If not defined, it reloads all heuristically matching layers.

    // MODEL PARAMETERS
    "encoder_type": "rel_pos_transformer",

    // TRAINING
    "batch_size": 128,       // Batch size for training. Lower values than 32 might cause hard to learn attention. It is overwritten by 'gradual_training'.
    "eval_batch_size":64,
    "r": 1,                 // Number of decoder frames to predict per iteration. Set the initial values if gradual training is enabled.
    "loss_masking": true,   // enable / disable loss masking against the sequence padding.

    // VALIDATION
    "run_eval": true,
    "test_delay_epochs": 1,       //Until attention is aligned, testing only wastes computation time.
    "test_sentences_file": null,  // set a file to load sentences to be used for testing. If it is null then we use default english sentences.
    "test_sentences": [
        "早安，你好呀!"
    ],
    // OPTIMIZER
    "noam_schedule": true,         // use noam warmup and lr schedule.
    "grad_clip": 5.0,              // upper limit for gradients for clipping.
    "epochs": 5,               // total number of epochs to train.
    "lr": 1e-3,                    // Initial learning rate. If Noam decay is active, maximum learning rate.
    "wd": 0.000001,                // Weight decay weight.
    "warmup_steps": 4000,          // Noam decay steps to increase the learning rate from 0 to "lr"
    "seq_len_norm": false,         // Normalize eash sample loss with its length to alleviate imbalanced datasets. Use it if your dataset is small or has skewed distribution of sequence lengths.


    // TENSORBOARD and LOGGING
    "print_step": 25,       // Number of steps to log training on console.
    "tb_plot_step": 100,    // Number of steps to plot TB training figures.
    "print_eval": false,     // If True, it prints intermediate loss values in evalulation.
    "save_step": 5000,      // Number of training steps expected to save traninpg stats and checkpoints.
    "checkpoint": true,     // If true, it saves checkpoints per "save_step"
    "tb_model_param_stats": false,     // true, plots param stats per layer on tensorboard. Might be memory consuming, but good for debugging.
    "apex_amp_level": null,

    // DATA LOADING
    "text_cleaner": "chinese_mandarin_cleaners",
    "enable_eos_bos_chars": false, // enable/disable beginning of sentence and end of sentence chars.
    "num_loader_workers": 8,        // number of training data loader processes. Don't set it too big. 4-8 are good values.
    "num_val_loader_workers": 8,    // number of evaluation data loader processes.
    "batch_group_size": 0,  //Number of batches to shuffle after bucketing.
    "min_seq_len": 2,       // DATASET-RELATED: minimum text length to use in training
    "max_seq_len": 500,     // DATASET-RELATED: maximum text length
    "compute_f0": false,     // compute f0 values in data-loader

    // PATHS
    "output_path": "./",

    // PHONEMES
    "phoneme_cache_path": "./phoneme_cache",  // phoneme computation is slow, therefore, it caches results in the given folder.
    "use_phonemes": true,
    "use_espeak_phonemes": true,// use phonemes instead of raw characters. It is suggested for better pronounciation.
    "phoneme_language": "zh-CN",     // depending on your target language, pick one from  https://github.com/bootphon/phonemizer#languages

    // MULTI-SPEAKER and GST
    "use_speaker_embedding": true,     // use speaker embedding to enable multi-speaker learning.
    "use_d_vector_file": true, // if true, forces the model to use external embedding per sample instead of nn.embeddings, that is, it supports external embeddings such as those used at: https://arxiv.org/abs /1806.04558
    "d_vector_file": "./embedding/speakers.json", // if not null and use_external_speaker_embedding_file is true, it is used to load a specific embedding file and thus uses these embeddings instead of nn.embeddings, that is, it supports external embeddings such as those used at: https://arxiv.org/abs /1806.04558
    "d_vector_dim": 512,

    "style_wav_for_test": null,          // path to style wav file to be used in TacotronGST inference.
    "use_gst": false,       // TACOTRON ONLY: use global style tokens

    "stopnet": false,
    "separate_stopnet": false,
    
    // DATASETS
    "datasets":   // List of datasets. They all merged and they get different speaker_ids.
        [
            {
                "name": "aishell3",
                "path": "./dataset/aishell_3/",
                "meta_file_train": "t_metadata.csv",
                "meta_file_val": null
            }
        ]
}

**Reply by erogol** [reply]
That still doesn't help to see what is wrong. You either need to provide something reproducible or just debug the core yourself at the line that causes the error.

**Reply by Lizh-TW** [reply]
@erogol, I have created a Colab.
https://colab.research.google.com/drive/1ZJGBYi_Em12obIAeQncEteum4GkVrXIb?usp=sharing
Please tell me how to solve this problem, thank you.

**Comment by Lizh-TW** [comment, marked as answer]
I found a possible problem.
![image](https://user-images.githubusercontent.com/58511222/132140276-29694461-1474-41a0-8904-8c983adb1660.png)
[1, 450, 512] "450" is the amount of the speaker owner voice file.
![image](https://user-images.githubusercontent.com/58511222/132140342-67d71639-876d-43e4-896f-0a58a02fcb8a.png)
Maybe embedding one voice file at a time?

**Reply by Lizh-TW** [reply]
Adding this line in tacotron2.py inference
![image](https://user-images.githubusercontent.com/58511222/132284801-0000c326-eb76-477b-96e6-f21fc24064f2.png)
